build_tools/upload_logs_to_s3.py

```diff
@@ -2,45 +2,63 @@
 """
 upload_logs_to_s3.py
 
-Uploads log files and index.html to an S3 bucket using the AWS CLI.
+Uploads log files and index.html to an S3 bucket using boto3.
 """
 
 import os
 import sys
-import glob
-import shutil
 import argparse
-import subprocess
 from pathlib import Path
 
+import boto3
+from botocore.exceptions import ClientError
+
 
 def log(*args, **kwargs):
     print(*args, **kwargs)
     sys.stdout.flush()
 
 
-def check_aws_cli_available():
-    if not shutil.which("aws"):
-        log("[ERROR] AWS CLI not found in PATH.")
-        sys.exit(1)
-
+def upload_file_boto3(file_path: Path, bucket: str, key: str, content_type: str = None):
+    s3 = boto3.client("s3")
+    extra_args = {"ContentType": content_type} if content_type else {}
 
-def run_aws_cp(source_path: Path, s3_destination: str, content_type: str = None):
-    if source_path.is_dir():
-        cmd = ["aws", "s3", "cp", str(source_path), s3_destination, "--recursive"]
-    else:
-        cmd = ["aws", "s3", "cp", str(source_path), s3_destination]
+    # Skip empty files (avoids IncompleteBody errors)
+    if not file_path.exists() or file_path.stat().st_size == 0:
+        log(f"[WARN] Skipping upload of empty or missing file: {file_path}")
+        return
 
-    if content_type:
-        cmd += ["--content-type", content_type]
     try:
-        log(f"[INFO] Running: {' '.join(cmd)}")
-        subprocess.run(cmd, check=True)
-    except subprocess.CalledProcessError as e:
-        log(f"[ERROR] Failed to upload {source_path} to {s3_destination}: {e}")
+        # Check if file already exists
+        s3.head_object(Bucket=bucket, Key=key)
+        already_exists = True
+    except ClientError as e:
+        if e.response["Error"]["Code"] == "404":
+            already_exists = False
+        else:
+            log(f"[ERROR] Failed checking existence for {key}: {e}")
+            return
+
+    if already_exists:
+        # Upload anyway for the additional logs,
+        # but do not repeat success log
+        try:
+            s3.upload_file(str(file_path), bucket, key, ExtraArgs=extra_args)
+        except ClientError as e:
+            log(f"[ERROR] Re-upload failed for {file_path}: {e}")
+        return  # Suppress success log
+
+    # First-time upload
+    try:
+        log(f"[INFO] Uploading {file_path} to s3://{bucket}/{key}")
+        s3.upload_file(str(file_path), bucket, key, ExtraArgs=extra_args)
+    except ClientError as e:
+        log(f"[ERROR] Failed to upload {file_path} to s3://{bucket}/{key}: {e}")
+    else:
+        log(f"[INFO] Successfully uploaded {file_path} to s3://{bucket}/{key}")
 
 
-def upload_logs_to_s3(s3_base_path: str, build_dir: Path):
+def upload_logs_to_s3(bucket: str, subdir: str, build_dir: Path):
     log_dir = build_dir / "logs"
```

Comment:
Here also, take `--logs-dir` instead of `--build-dir`

Reply:
Updated